File: frontend/src/services/catalog.ts
"use client";

import api from './api';

export interface SubjectSummary {
    id: string;
    name: string;
    description?: string | null;
}

export interface SubjectCreatePayload {
    name: string;
    description?: string | null;
}

export interface SubjectUpdatePayload {
    name?: string;
    description?: string | null;
}

export interface TermSummary {
    id: string;
    code: string;
    name: string;
    status: string;
    academic_year_name?: string | null;
}

export interface StudentSummary {
    id: string;
    user_id?: string | null;
    subject_id?: string | null;
    email?: string | null;
    full_name?: string | null;
}

export interface TopicSummary {
    id: string;
    subject_id: string;
    term_id?: string | null;
    code?: string | null;
    name: string;
}

export async function fetchSubjects(mine = true): Promise<SubjectSummary[]> {
    const res = await api.get('/catalog/subjects', { params: { mine } });
    return res.data as SubjectSummary[];
}

export async function createSubject(payload: SubjectCreatePayload): Promise<SubjectSummary> {
    const res = await api.post('/catalog/subjects', payload);
    return res.data as SubjectSummary;
}

export async function updateSubject(
    subjectId: string,
    payload: SubjectUpdatePayload,
): Promise<SubjectSummary> {
    const res = await api.put(`/catalog/subjects/${subjectId}`, payload);
    return res.data as SubjectSummary;
}

export async function deleteSubject(subjectId: string): Promise<void> {
    await api.delete(`/catalog/subjects/${subjectId}?force=true`);
}

export async function assignStudentSubject(studentId: string, subjectId: string): Promise<StudentSummary> {
    const res = await api.patch(`/catalog/students/${studentId}`, { subject_id: subjectId });
    return res.data as StudentSummary;
}

export async function fetchTerms(active: boolean | null = true): Promise<TermSummary[]> {
    const res = await api.get('/catalog/terms', { params: { active } });
    return res.data as TermSummary[];
}

export async function fetchStudents(mine = true, subjectId?: string): Promise<StudentSummary[]> {
    const res = await api.get('/catalog/students', {
        params: { mine, subject_id: subjectId },
    });
    return res.data as StudentSummary[];
}

export async function fetchTopics(params: {
    mine?: boolean;
    subjectId?: string;
    termId?: string;
}): Promise<TopicSummary[]> {
    const res = await api.get('/catalog/topics', {
        params: {
            mine: params.mine ?? true,
            subject_id: params.subjectId,
            term_id: params.termId,
        },
    });
    return res.data as TopicSummary[];
}

File: backend/app/schemas/catalog.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, constr


class SubjectSummary(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None

    model_config = ConfigDict(from_attributes=True)


class SubjectCreate(BaseModel):
    name: constr(strip_whitespace=True, min_length=1, max_length=255)
    description: str | None = None


class SubjectUpdate(BaseModel):
    name: constr(strip_whitespace=True, min_length=1, max_length=255) | None = None
    description: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TermSummary(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    status: str
    academic_year_name: str | None = None
    start_date: date | None = None
    end_date: date | None = None

    model_config = ConfigDict(from_attributes=True)


class StudentSummary(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID | None = None
    subject_id: uuid.UUID | None = None
    enrollment_date: datetime | None = None
    email: str | None = None
    full_name: str | None = None


class StudentSubjectUpdate(BaseModel):
    subject_id: uuid.UUID

    model_config = ConfigDict(from_attributes=True)


class TopicSummary(BaseModel):
    id: uuid.UUID
    subject_id: uuid.UUID
    term_id: uuid.UUID | None = None
    code: str | None = None
    name: str

    model_config = ConfigDict(from_attributes=True)
